src/App.tsx

```diff
@@ -14,11 +17,160 @@ const goodsFromServer: string[] = [
   'Garlic',
 ];
 
-const App: React.FC = () => (
-  <div className="App">
-    <h1>Goods</h1>
-    {goodsFromServer.length}
-  </div>
-);
+interface State {
+  initialGoods: string[];
+  isStarted: boolean;
+  isReversed: boolean;
+  sortBy: string;
+  minLength: number;
+}
+
+class App extends React.PureComponent<{}, State> {
+  state: State = {
+    initialGoods: goodsFromServer,
+    isStarted: true,
+    isReversed: false,
+    sortBy: 'reset',
+    minLength: 1,
+  };
+
+  onStartButtonClick = () => {
+    this.setState({
+      isStarted: false,
+    });
+  };
+
+  onReverseButtonClick = () => {
+    this.setState(state => ({
+      isReversed: !state.isReversed,
+    }));
+  };
+
+  onSortByNameButtonClick = () => {
+    this.setState({
+      isReversed: false,
+      sortBy: 'name',
+    });
+  };
+
+  onSortByLengthButtonClick = () => {
+    this.setState({
+      isReversed: false,
+      sortBy: 'length',
+    });
+  };
+
+  onResetButtonClick = () => {
+    this.setState({
+      minLength: 1,
+      isReversed: false,
+      sortBy: 'reset',
+    });
+  };
+
+  render() {
+    const {
+      initialGoods,
+      isStarted,
+      isReversed,
+      sortBy,
+      minLength,
+    } = this.state;
+
+    const currentGoods = [...initialGoods]
+      .filter((product) => product.length >= minLength);
+
+    currentGoods.sort((firstProduct, secondProduct) => {
+      switch (sortBy) {
+        case 'name':
+          return firstProduct.localeCompare(secondProduct);
+
+        case 'length':
+          return firstProduct.length - secondProduct.length;
+
+        default:
+          return 0;
+      }
+    });
+
+    if (isReversed) {
+      currentGoods.reverse();
+    }
```

Comment:
This logic shouldn't be here. Method `render` should be as clear as possible.
Let's move this logic to a separate method